File: core/src/main/java/com/adamblakey/uptimererer/probe/Probe.java
package com.adamblakey.uptimererer.probe;

import com.adamblakey.uptimererer.events.SiteConfig;
import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.time.Instant;

/**
 * Performs a single HTTP(S) uptime check against a site.
 *
 * <p>The result is {@code up} iff the request completes and the response
 * status is one of the site's expected codes. TLS certificates are validated
 * (the JDK client's default) and redirects are followed up to the JDK's limit
 * ({@code jdk.httpclient.redirects.retrylimit}, default 5); a redirect that
 * never resolves therefore surfaces as a non-expected status rather than a
 * hang.
 */
public final class Probe {

    /** Runs one request against the site using its configured method and timeout. */
    public ProbeResult check(SiteConfig site) {
        Duration timeout = Duration.ofSeconds(site.timeoutSeconds());

        final HttpRequest request;
        try {
            request = HttpRequest.newBuilder()
                    .uri(URI.create(site.url()))
                    .timeout(timeout)
                    .method(site.method(), HttpRequest.BodyPublishers.noBody())
                    .build();
        } catch (RuntimeException e) {
            return new ProbeResult(false, 0, Duration.ZERO, "build request: " + describe(e));
        }

        Instant start = Instant.now();
        try (HttpClient client = HttpClient.newBuilder()
                .followRedirects(HttpClient.Redirect.NORMAL)
                .connectTimeout(timeout)
                .build()) {

            HttpResponse<Void> response = client.send(request, HttpResponse.BodyHandlers.discarding());
            Duration latency = Duration.between(start, Instant.now());
            int status = response.statusCode();
            if (site.statusExpected(status)) {
                return new ProbeResult(true, status, latency, "");
            }
            return new ProbeResult(false, status, latency, "unexpected status " + status);
        } catch (IOException e) {
            return new ProbeResult(false, 0, Duration.between(start, Instant.now()), describe(e));
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            return new ProbeResult(false, 0, Duration.between(start, Instant.now()), describe(e));
        }
    }

    private static String describe(Exception e) {
        String message = e.getMessage();
        return (message == null || message.isBlank()) ? e.getClass().getSimpleName() : message;
    }
}

File: internal/probe/probe.go
// Package probe performs a single HTTP(S) uptime check against a site.
package probe

import (
	"context"
	"fmt"
	"io"
	"net/http"
	"time"

	"github.com/adamblakey/uptimererer/internal/events"
)

const maxRedirects = 5

// Result is the outcome of one probe.
type Result struct {
	Up         bool
	HTTPStatus int           // 0 when no response was received
	Latency    time.Duration // time to (failed) response
	Err        string        // empty when Up
}

// Check performs one request against the site using its configured method and
// timeout. The result is Up iff the request completes and the response status
// is one of the site's expected codes. TLS is validated and redirects are
// followed up to a small limit.
func Check(ctx context.Context, site events.SiteConfig) Result {
	client := &http.Client{
		Timeout: time.Duration(site.TimeoutSeconds) * time.Second,
		CheckRedirect: func(req *http.Request, via []*http.Request) error {
			if len(via) >= maxRedirects {
				return fmt.Errorf("stopped after %d redirects", maxRedirects)
			}
			return nil
		},
	}

	req, err := http.NewRequestWithContext(ctx, site.Method, site.URL, nil)
	if err != nil {
		return Result{Err: fmt.Sprintf("build request: %v", err)}
	}

	start := time.Now()
	resp, err := client.Do(req)
	latency := time.Since(start)
	if err != nil {
		return Result{Latency: latency, Err: err.Error()}
	}
	defer resp.Body.Close()
	_, _ = io.Copy(io.Discard, io.LimitReader(resp.Body, 64<<10))

	res := Result{HTTPStatus: resp.StatusCode, Latency: latency}
	if site.StatusExpected(resp.StatusCode) {
		res.Up = true
	} else {
		res.Err = fmt.Sprintf("unexpected status %d", resp.StatusCode)
	}
	return res
}
